rp2040-hal/src/gpio.rs

```diff
@@ -0,0 +1,132 @@
+//! General Purpose Input and Output (GPIO)
+//!
+//! TODO: Docs for this whole module
+//!
+//! For example, to turn on the LED on a Pico board:
+//!
+//! ```rust
+//! let mut pac = rp2040_pac::Peripherals::take().unwrap();
+//!
+//! let pins = pac.IO_BANK0.split(&mut pac.RESETS);
+//! let mut led_pin = pins.gpio25.into_output(&mut pac.PADS_BANK0, &mut pac.SIO);
+//!  led_pin.set_low().unwrap();
+//! ```
+use core::convert::Infallible;
+use core::marker::PhantomData;
+use embedded_hal::digital::v2::OutputPin;
+
+pub struct Floating;
+pub struct Input<MODE> {
+    _mode: PhantomData<MODE>,
+}
+pub struct Output;
+
+pub trait GpioExt {
+    type Parts;
+
+    // TODO: Do we need a marker to check that clocks are up?
+    fn split(self, reset: &mut rp2040_pac::RESETS) -> Self::Parts;
+}
+
+const FUNCTION_SIO: u8 = 5;
+
+macro_rules! gpio {
+    ($GPIOX:ident, $gpiox:ident, $PADSX:ident, [
+        $($PXi:ident: ($pxi:ident, $i:expr),)+
+    ]) => {
+        impl GpioExt for pac::$GPIOX {
+            type Parts = Parts;
+
+            fn split(self, resets: &mut pac::RESETS) -> Parts {
+                resets.reset.modify(|_, w| w.$gpiox().clear_bit());
+                Parts {
+                    $(
+                        $pxi: $PXi { _mode: PhantomData },
+                    )+
+                }
+            }
+        }
+
+        pub struct Parts {
+            $(
+                pub $pxi: $PXi<Input<Floating>>,
+            )+
+        }
+
+        $(
+            pub struct $PXi<MODE> {
+                _mode: PhantomData<MODE>,
+            }
+
+            impl<MODE> $PXi<MODE> {
+                pub fn into_output(
+                    self,
+                    pads: &mut pac::$PADSX,
+                    sio: &mut pac::SIO,
+                ) -> $PXi<Output> {
+                    pads.gpio[$i].modify(|_, w| w.ie().set_bit().od().clear_bit());
+                    unsafe {
+                        (*pac::$GPIOX::ptr()).gpio[$i]
+                            .gpio_ctrl
+                            .write_with_zero(|x| x.funcsel().bits(FUNCTION_SIO));
+                    }
+                    sio.gpio_oe.modify(|_, x| unsafe { x.bits(1 << $i) });
```

Comment:
If you move the ownership of the SIO into this driver you should use the atomic version of this register `GPIO_OE_SET` with `.write()` instead of `.modify()`.

Then you can also pass around pins without giving the SIO instance around with it.